odoo/orm/domains.py

```diff
@@ -894,8 +897,30 @@ def _optimize_for_sql(self, model: BaseModel) -> Domain:
             if not field.search:
                 _logger.error("Non-stored field %s cannot be searched.", field, stack_info=_logger.isEnabledFor(logging.DEBUG))
                 return _TRUE_DOMAIN
-            computed_domain = field.determine_domain(model, self.operator, self.value)
-            return Domain(computed_domain)._optimize_for_sql(model)
+            operator, value = self.operator, self.value
+            # simplify further for boolean fields
+            if field.type == 'boolean':
+                # special case to ease implementation: always use (field_expr, 'in'/'not in', [True])
+                if not (operator in ('in', 'not in') and isinstance(value, COLLECTION_TYPES) and len(value) == 1):
+                    self._raise("Cannot compare %r to %s which is not a collection of length 1", self.field, type(value))
+                [truth] = value
+                if not truth:
+                    operator = _INVERSE_OPERATOR[operator]
+                value = [True]
+            try:
+                # use the `search` function of the field
+                computed_domain = Domain(field.determine_domain(model, operator, value))
+            except NotImplementedError as e:
+                # when not supported, check if it is supported for the positive operator
+                computed_domain = None
+                if positive_operator := NEGATIVE_CONDITION_OPERATORS.get(operator):
+                    try:
+                        computed_domain = ~Domain(field.determine_domain(model, positive_operator, value))
+                    except NotImplementedError:
+                        computed_domain = None
+                if computed_domain is None:
+                    raise NotImplementedError(f"Unsupported operator in {self!r} on model {model._name}") from e
```

Comment:
Should we transform it into a translated UserError instead? With the domain selector, it is likely to happen on some searchable fields.

Reply:
We never raised UserError from domains, usually it is a ValueError or TypeError. So I would keep a technical error here as well. I could be another PR to review if we need some user-friendly errors (and we would need to translate them too).

Reply:
> We never raised UserError from domains, usually it is a ValueError or TypeError.

But we have in a lot of field search method, and you replace them with `NotImplementedError` :thinking: 
Then you actually change tons of user errors into programming errors. It will generate a tons of ticket because now the Custom Filter may return tracebacks instead of a user friendly error.

Reply:
Updated.